app/models/integration_customers/base_customer.rb

```diff
@@ -47,6 +53,21 @@ def self.customer_type(type)
         raise(NotImplementedError)
       end
     end
+
+    def tax_integration?
+      TAX_INTEGRATION_TYPES.include?(type)
+    end
+
+    private
+
+    def only_one_tax_integration_per_customer
+      conflict = IntegrationCustomers::BaseCustomer.where(customer_id:, type: TAX_INTEGRATION_TYPES)
+      conflict = conflict.where.not(id:) if persisted?
+
+      return unless conflict.exists?
+
+      errors.add(:type, "tax_integration_exists")
```

Comment:
Let's split it in smaller parts?

We can extract `.where(type: TAX_INTEGRATION_TYPES)` into scope `tax_kind` similar to already existing `accounting_kind`.
`where.not(id:)` could be replaced with `.excluding(self)`
and finally we can combine `.where(customer_id:` with `exists?` -> `.exists?(customer_id:)`.

```suggestion
      if IntegrationCustomers::BaseCustomer.tax_kind.excluding(self).exists?(customer_id:)
        errors.add(:type, "tax_integration_exists")
      end
```

Also it looks like tests for this validation does not check for this `conflict = conflict.where.not(id:) if persisted?` case. If you comment that line all tests will be green.

Reply:
added new scope but left old style which is more close to me:)